### PR TITLE
[PPP-5004]-Tomcat Configuration vulnerable to Server-Side Request Forgery

### PR DESCRIPTION
**Pull request by SahithiKommagani6**
[PPP-5004]-Tomcat Configuration vulnerable to Server-Side Request Forgery

[PPP-5004]: https://hv-eng.atlassian.net/browse/PPP-5004?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 20m 33s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions
```
---
 :exclamation: No tests found!
 ---
 
 

:information_source:  _This is an automatic message_

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/system.properties [hunk 1]
```diff
@@ -77,3 +77,8 @@ cors-requests-allow-credentials=true
 
 # Exposed response HTTP headers for all cross-origin requests:
 cors-requests-exposed-headers=
+
+# ssrf protection  should be enabled by default for all requests to prevent server side request forgery attacks
+# Set to true to enable SSRF protection for the server.
+# ssrf protection is applied only if the ssrf-protection-enabled  is set to true
```
Comment (dcleao) [inline assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/system.properties:83]: I suggest the following rewording, in part for consistency with CSRF and CORS corresponding properties:
```properties
## SSRF ##

# Enables SSRF (Server-Side Request Forgery) protection for the server.
#
# Set to `true` to enable SSRF protection for the server.
#
# When protection is enabled, the server will only accept HTTP requests whose Host header
# is compatible with one of the server's fully-qualified URLs, as configured in `fully-qualified-server-url` 
# and `alternate-fully-qualified-server-urls`, in `server.properties`. 
# This applies to all endpoints exposed by the server.
# 
# *ATTENTION*:
# Setting this option to false opens the server and all of its plugins to SSRF attacks.
```

Additionally, given that SSRF is the foundation for CORS/CSRF protection, I propose placing the SSRF section before the CSRF one.

**Review by dcleao — CHANGES_REQUESTED** [review]

**Comment by SahithiKommagani6** [comment]
https://github.com/pentaho/security-web/pull/28

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security.xml [hunk 2]
```diff
@@ -98,7 +98,8 @@
                                    filterInvocationInterceptorForWS" />
 
         <sec:filter-chain pattern="/**"
-                          filters="csrfGateFilter,
+                          filters="ssrfGateFilter,
```
[the remark this answers is not in the document]
Comment (SahithiKommagani6) [reply assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security.xml:107]: addressed the review comment
  <sec:filter-chain pattern="/api/csrf/token"
                          filters="ssrfGateFilter,
                                   csrfGateFilter,
                                   corsFilter" />

        <sec:filter-chain pattern="/webservices/**"
                          filters="ssrfGateFilter,

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/server.properties [hunk 3]
```diff
@@ -2,6 +2,11 @@
 # and whenever something need to talk back to the server
 fully-qualified-server-url=http://localhost:8080/pentaho/
 
+# AlternateFullyQualifiedServerUrl is used only in the case when ssrf protection is enabled
```
[the remark this answers is not in the document]
Comment (SahithiKommagani6) [reply assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/server.properties]: modified documentation related to urls
# The main fully-qualified URL by which the server can be reached.
#
# Example: http://my-company.com:8080/pentaho/
#
# Alternate addresses can be configured in `alternate-fully-qualified-server-urls`.
# Currently, this property is used for generating content that points back to the server,
# either offline content, or online content for same-frame embedding scenarios.
# Additionally, this property is used for protection against SSRF attacks (via spoofing of the HTTP request's Host header), limiting the allowed Hosts. To enable SSRF protection, set the property `ssrf-protection-enabled` to `true`, in `system.properties`.

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/system.properties (hunk 1 above)
Comment (SahithiKommagani6) [reply assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/system.properties]: modified documentation

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/server.properties [hunk 4]
```diff
@@ -2,6 +2,11 @@
...
+# and the server is behind a proxy
+# can be used to specify a different server url to be used in the case of ssrf protection
+alternate-fully-qualified-server-urls=http://172.16.13.216:8080,http://172.16.13.38:8080,http://172.16.11.83
```
[the remark this answers is not in the document]
Comment (SahithiKommagani6) [reply assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/server.properties]: property set to empty by default
alternate-fully-qualified-server-urls=

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/system.properties [hunk 5]
```diff
@@ -77,3 +77,8 @@ cors-requests-allow-credentials=true
...
+ssrf-protection-enabled=true
```
[the remark this answers is not in the document]
Comment (SahithiKommagani6) [reply assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/system.properties]: included new line at the end of file

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 15m 18s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions
```
---
 :exclamation: No tests found!
 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 10m 26s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions
```
---
 :exclamation: No tests found!
 ---
 
 

:information_source:  _This is an automatic message_

**Review by dcleao — CHANGES_REQUESTED** [review]

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/server.properties [hunk 6]
```diff
@@ -1,7 +1,25 @@
-# FullyQualifiedServerUrl is used only in the case of offline content generation
-# and whenever something need to talk back to the server
+# The main fully-qualified URL by which the server can be reached.
+#
+# Example: http://my-company.com:8080/pentaho/
+#
+# Alternate addresses can be configured in `alternate-fully-qualified-server-urls`.
```
[the remark this answers is not in the document]
Comment (SahithiKommagani6) [reply assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/server.properties]: Addressed

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 18m 3s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions
```
---
 :exclamation: No tests found!
 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 8m 46s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions
```
---
 :exclamation: No tests found!
 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :x: Build failed in 6m 1s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,extensions
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :x: Build failed in 3m 8s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,extensions
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

File: extensions/pom.xml [hunk 7]
```diff
@@ -2083,14 +2083,17 @@
     <dependency>
       <groupId>com.hitachivantara.security.web</groupId>
       <artifactId>security-web-server-api</artifactId>
+      <version>0.9.0</version>
```
Comment (dcleao) [inline extensions/pom.xml:2086]: Version should be changed in the global pentaho parent pom:
https://github.com/pentaho/maven-parent-poms/blob/4a6d77cda21945af9aa382f177ccd7eab53507c1/pom.xml#L373

File: extensions/src/main/java/org/pentaho/platform/config/PentahoSsrfConfigurationPojo.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.Set;
+
+public class PentahoSsrfConfigurationPojo extends SsrfConfigurationPojo {
+
+    public PentahoSsrfConfigurationPojo ( URI mainFullyQualifiedUrl, Set<URI> alternativeFullyQualifiedUrls,
+                                         boolean isEnabled ) {
+        super ( baseUrls ( mainFullyQualifiedUrl, alternativeFullyQualifiedUrls ), isEnabled );
+    }
+
+    private static Set<URI> baseUrls ( URI mainFullyQualifiedUrl, Set<URI> alternativeFullyQualifiedUrls ) {
```
Comment (dcleao) [inline extensions/src/main/java/org/pentaho/platform/config/PentahoSsrfConfigurationPojo.java:36]: Perhaps, better named as `buildAllowedOrigins`.

File: extensions/src/main/java/org/pentaho/platform/config/PentahoSsrfConfigurationPojo.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        Set<URI> baseUrls = new HashSet<> ( alternativeFullyQualifiedUrls );
+        baseUrls.add ( mainFullyQualifiedUrl );
+        return baseUrls;
```
Comment (dcleao) [inline extensions/src/main/java/org/pentaho/platform/config/PentahoSsrfConfigurationPojo.java:39]: I would prefer if we would try to preserve the order of the URLs, which may lead to some performance improvements.

Something like this should work:

```java
public PentahoSsrfConfigurationPojo ( URI mainFullyQualifiedUrl, 
                 List<URI> alternativeFullyQualifiedUrls,
                 boolean isEnabled ) {
       super ( buildAllowedOrigins( mainFullyQualifiedUrl, alternativeFullyQualifiedUrls ), isEnabled );
    }

    private static Set<URI> buildAllowedOrigins( URI mainFullyQualifiedUrl, List<URI> alternativeFullyQualifiedUrls ) {
      Set<URI> allowedOrigins = new LinkedHashSet<>();
      allowedOrigins.add ( mainFullyQualifiedUrl );
      allowedOrigins.addAll( alternativeFullyQualifiedUrls );
      return allowedOrigins;
   }
```

**Review by dcleao — CHANGES_REQUESTED** [review]

**Comment by buildguy** [comment]
### :x: Build failed in 3m 23s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,extensions
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :x: Build failed in 3m 27s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,extensions
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :x: Build failed in 4m 39s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,extensions
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :x: Build failed in 4m 12s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,extensions
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security-ssrf.xml [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    <bean id="pentahoSsrfConfigurationPojo" class="org.pentaho.platform.config.PentahoSsrfConfigurationPojo">
+        <constructor-arg value="${server.fully-qualified-server-url}"/>
+        <constructor-arg value="${server.alternative-fully-qualified-server-urls}"/>
+        <constructor-arg value= "${system.ssrf-protection-enabled:false}"/>
+    </bean>
+
+    <!--servlet filter class that will be used to filter the request and response for SSRF protection -->
+    <!--the ssrfConfig and ssrfUtil beans are injected into the filter
+     ssrf protection is applied only if the ssrfEnabled is set to true
+     the filter will check if the request is a valid request or not -->
```
Comment (dcleao) [inline assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security-ssrf.xml:23]: This comment seems to not add much. I would remove it.

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security-ssrf.xml [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    <!-- read the alternate-fully-qualified-server-url and fully-qualified-server-url from the system.properties file -->
+    <!--  read the ssrf-protection-enabled from the system.properties file -->
+
+    <bean id="pentahoSsrfConfigurationPojo" class="org.pentaho.platform.config.PentahoSsrfConfigurationPojo">
+        <constructor-arg value="${server.fully-qualified-server-url}"/>
+        <constructor-arg value="${server.alternative-fully-qualified-server-urls}"/>
+        <constructor-arg value= "${system.ssrf-protection-enabled:false}"/>
+    </bean>
+
+    <!--servlet filter class that will be used to filter the request and response for SSRF protection -->
```
Comment (dcleao) [inline assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security-ssrf.xml:20]: Please add a space between: `<!-- servlet filter`.

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security-ssrf.xml [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+  | Application context containing SSRF related beans.
+  +-->
+<beans xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xmlns="http://www.springframework.org/schema/beans"
+       xsi:schemaLocation="
+          http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-4.3.xsd"
+       default-lazy-init="true">
+
+    <!-- read the alternate-fully-qualified-server-url and fully-qualified-server-url from the system.properties file -->
+    <!--  read the ssrf-protection-enabled from the system.properties file -->
```
Comment (dcleao) [inline assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security-ssrf.xml:12]: Please add a single space between: `<!-- read`

**Review by dcleao — CHANGES_REQUESTED** [review]

**Comment by buildguy** [comment]
### :x: Build failed in 4m 32s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,extensions
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security-ssrf.xml [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+<!--+
+  | Application context containing SSRF related beans.
+  +-->
+<beans xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xmlns="http://www.springframework.org/schema/beans"
+       xsi:schemaLocation="
+          http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-4.3.xsd"
+       default-lazy-init="true">
+
+    <!-- read the alternate-fully-qualified-server-url and fully-qualified-server-url from the system.properties file -->
```
Comment (dcleao) [inline assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security-ssrf.xml:11]: Should be `alternative-fully-qualified-server-url`.

File: extensions/pom.xml [hunk 14]
```diff
@@ -805,7 +805,7 @@
         </exclusion>
       </exclusions>
     </dependency>
-    
+
```
Comment (dcleao) [inline extensions/pom.xml:808]: File has no (non-whitespace) changes and so should not be part of the PR.

**Review by dcleao — CHANGES_REQUESTED** [review]

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security-ssrf.xml (hunk 13 above)
Comment (SahithiKommagani6) [reply assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security-ssrf.xml]: addressed

**Comment by buildguy** [comment]
### :x: Build failed in 3m 22s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,extensions
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :x: Build failed in 3m 13s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,extensions
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :x: Build failed in 2m 37s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,extensions
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :x: Build failed in 3m 45s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,extensions
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

File: extensions/pom.xml (hunk 14 above)
Comment (SahithiKommagani6) [reply extensions/pom.xml]: addressed

**Review by dcleao — APPROVED** [review]

**Comment by buildguy** [comment]
### :x: Build failed in 3m 30s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
assemblies/pentaho-solutions,extensions
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_
